File: src/main/java/net/flyingfishflash/loremlist/api/LrmListApiService.java
package net.flyingfishflash.loremlist.api;

import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import net.flyingfishflash.loremlist.api.data.request.LrmItemCreateRequest;
import net.flyingfishflash.loremlist.api.data.request.LrmListCreateRequest;
import net.flyingfishflash.loremlist.api.data.response.ApiServiceResponse;
import net.flyingfishflash.loremlist.api.data.response.AssociationDeletedResponse;
import net.flyingfishflash.loremlist.api.data.response.AssociationsDeletedResponse;
import net.flyingfishflash.loremlist.api.data.response.LrmItemResponse;
import net.flyingfishflash.loremlist.api.data.response.LrmListDeletedResponse;
import net.flyingfishflash.loremlist.api.data.response.LrmListItemAddedResponse;
import net.flyingfishflash.loremlist.api.data.response.LrmListItemMovedResponse;
import net.flyingfishflash.loremlist.api.data.response.LrmListItemResponse;
import net.flyingfishflash.loremlist.api.data.response.LrmListResponse;
import net.flyingfishflash.loremlist.core.response.structure.ApiMessageNumeric;

public interface LrmListApiService {
  // list context scoped by owner
  ApiServiceResponse<ApiMessageNumeric> countByOwner(String owner);

  ApiServiceResponse<LrmListResponse> create(
      LrmListCreateRequest lrmListCreateRequest, String owner);

  ApiServiceResponse<LrmListDeletedResponse> deleteByOwner(String owner);

  ApiServiceResponse<LrmListDeletedResponse> deleteByOwnerAndId(
      UUID id, String owner, boolean removeItemAssociations);

  ApiServiceResponse<List<LrmItemResponse>> findEligibleItemsByOwner(UUID listId, String owner);

  ApiServiceResponse<List<LrmListResponse>> findByOwner(String owner);

  ApiServiceResponse<List<LrmListResponse>> findByOwnerExcludeItems(String owner);

  ApiServiceResponse<List<LrmListResponse>> findByOwnerAndHavingNoItemAssociations(String owner);

  ApiServiceResponse<LrmListResponse> findByOwnerAndId(UUID id, String owner);

  ApiServiceResponse<LrmListResponse> findByOwnerAndIdExcludeItems(UUID id, String owner);

  ApiServiceResponse<List<LrmListResponse>> findByPublic();

  ApiServiceResponse<List<LrmListResponse>> findByPublicExcludeItems();

  ApiServiceResponse<LrmListResponse> patchByOwnerAndId(
      UUID id, String owner, Map<String, Object> patchRequest);

  // list item context scoped by owner
  ApiServiceResponse<LrmListItemAddedResponse> addListItem(
      UUID listId, Set<UUID> itemIdCollection, String owner);

  ApiServiceResponse<ApiMessageNumeric> countListItems(UUID listId, String listOwner);

  ApiServiceResponse<LrmListItemResponse> createListItem(
      UUID listId, LrmItemCreateRequest itemCreateRequest, String creator);

  ApiServiceResponse<LrmListItemResponse> findListItem(UUID listId, UUID itemId, String listOwner);

  ApiServiceResponse<LrmListItemMovedResponse> moveListItem(
      UUID listId, UUID itemId, UUID destinationListId, String owner);

  ApiServiceResponse<AssociationDeletedResponse> removeListItem(
      UUID listId, UUID itemId, String componentsOwner);

  ApiServiceResponse<AssociationsDeletedResponse> removeAllListItems(UUID listId, String listOwner);

  ApiServiceResponse<LrmListItemResponse> patchListItem(
      UUID listId, UUID itemId, String listOwner, Map<String, Object> patchRequest);
}

File: src/main/kotlin/net/flyingfishflash/loremlist/api/LrmListApiService.kt
package net.flyingfishflash.loremlist.api

import net.flyingfishflash.loremlist.api.data.request.LrmItemCreateRequest
import net.flyingfishflash.loremlist.api.data.request.LrmListCreateRequest
import net.flyingfishflash.loremlist.api.data.response.ApiServiceResponse
import net.flyingfishflash.loremlist.api.data.response.AssociationDeletedResponse
import net.flyingfishflash.loremlist.api.data.response.AssociationsDeletedResponse
import net.flyingfishflash.loremlist.api.data.response.LrmItemResponse
import net.flyingfishflash.loremlist.api.data.response.LrmListDeletedResponse
import net.flyingfishflash.loremlist.api.data.response.LrmListItemAddedResponse
import net.flyingfishflash.loremlist.api.data.response.LrmListItemMovedResponse
import net.flyingfishflash.loremlist.api.data.response.LrmListItemResponse
import net.flyingfishflash.loremlist.api.data.response.LrmListResponse
import net.flyingfishflash.loremlist.core.response.structure.ApiMessageNumeric
import java.util.*

interface LrmListApiService {
  // list context scoped by owner
  fun countByOwner(owner: String): ApiServiceResponse<ApiMessageNumeric>
  fun create(lrmListCreateRequest: LrmListCreateRequest, owner: String): ApiServiceResponse<LrmListResponse>
  fun deleteByOwner(owner: String): ApiServiceResponse<LrmListDeletedResponse>
  fun deleteByOwnerAndId(id: UUID, owner: String, removeItemAssociations: Boolean): ApiServiceResponse<LrmListDeletedResponse>
  fun findEligibleItemsByOwner(listId: UUID, owner: String): ApiServiceResponse<List<LrmItemResponse>>
  fun findByOwner(owner: String): ApiServiceResponse<List<LrmListResponse>>
  fun findByOwnerExcludeItems(owner: String): ApiServiceResponse<List<LrmListResponse>>
  fun findByOwnerAndHavingNoItemAssociations(owner: String): ApiServiceResponse<List<LrmListResponse>>
  fun findByOwnerAndId(id: UUID, owner: String): ApiServiceResponse<LrmListResponse>
  fun findByOwnerAndIdExcludeItems(id: UUID, owner: String): ApiServiceResponse<LrmListResponse>
  fun findByPublic(): ApiServiceResponse<List<LrmListResponse>>
  fun findByPublicExcludeItems(): ApiServiceResponse<List<LrmListResponse>>
  fun patchByOwnerAndId(id: UUID, owner: String, patchRequest: Map<String, Any>): ApiServiceResponse<LrmListResponse>

  // list item context scoped by owner
  fun addListItem(listId: UUID, itemIdCollection: Set<UUID>, owner: String): ApiServiceResponse<LrmListItemAddedResponse>
  fun countListItems(listId: UUID, listOwner: String): ApiServiceResponse<ApiMessageNumeric>
  fun createListItem(listId: UUID, itemCreateRequest: LrmItemCreateRequest, creator: String): ApiServiceResponse<LrmListItemResponse>
  fun findListItem(listId: UUID, itemId: UUID, listOwner: String): ApiServiceResponse<LrmListItemResponse>
  fun moveListItem(listId: UUID, itemId: UUID, destinationListId: UUID, owner: String): ApiServiceResponse<LrmListItemMovedResponse>
  fun removeListItem(listId: UUID, itemId: UUID, componentsOwner: String): ApiServiceResponse<AssociationDeletedResponse>
  fun removeAllListItems(listId: UUID, listOwner: String): ApiServiceResponse<AssociationsDeletedResponse>
  fun patchListItem(listId: UUID, itemId: UUID, listOwner: String, patchRequest: Map<String, Any>): ApiServiceResponse<LrmListItemResponse>
}
